docs/reference-guide/modules/events/pages/event-versioning.adoc

```diff
@@ -3,45 +3,166 @@
+    private final int productCount;  // Computed from productIds
+
+    // Constructor that computes productCount
+    public EnrichedOrderPlacedEvent(String orderId, String customerId, List<String> productIds) {
+        this.orderId = orderId;
+        this.customerId = customerId;
+        this.productIds = productIds;
+        this.productCount = productIds != null ? productIds.size() : 0;
+    }
+
+    // getters...
+}
+----
+
+You can add a handler that receives the event in this new format:
+
+[source,java]
+----
+@EventHandler(eventName = "com.example.events.OrderPlaced")
+public void on(EnrichedOrderPlacedEvent event) {
+    // Axon converts the stored OrderPlacedEvent to EnrichedOrderPlacedEvent
+    // The productCount field is computed during conversion
+    notifyAnalytics(event.getProductCount());
+}
+----
+
+Both handlers process the same stored event, each receiving it in their preferred format.
+Axon's `Converter` performs the transformation automatically at handling time.
+
+For complete details on payload conversion, see xref:ROOT:conversion.adoc[Conversion].
+
+=== When payload conversion is sufficient
+
+Payload conversion handles many common versioning scenarios without requiring upcasters:
+
+* **Adding fields with defaults** - New fields can have default values during conversion
+* **Removing fields** - Handlers simply don't declare fields they don't need
+* **Renaming fields** - Use Jackson annotations like `@JsonProperty` to map old names to new fields
+* **Changing field types** - Converters can transform compatible types (for example String to Integer, Date to Instant)
+* **Restructuring for different handlers** - Different handlers can use completely different Java types for the same event
+
+=== When upcasters are needed
+
+Upcasters are required when you need to transform the **stored representation** of events in ways that payload conversion cannot handle:
+
+* **Splitting one event into multiple events** - When you need to split a single stored event into multiple distinct events
+* **Merging multiple events** - When you need to combine several stored events into one
+* **Complex structural transformations** - When the transformation requires access to multiple events or contextual information
+* **Changing event identity** - When you need to change the event's MessageType (qualified name or version) in storage
+
+For most scenarios, payload conversion is simpler and more flexible than upcasters.
+Only use upcasters when the stored event structure itself must change.
 
 [[event-upcasting]]
 == Event Upcasting
 
-Due to the ever-changing nature of software applications, it is likely that event definitions will also change over time.
-Since the Event Store is considered a read and append-only data source, your application must be able to read all events, regardless of when they were added.
-This is where upcasting comes in.
+[IMPORTANT]
+.Upcasters not available in Axon 5.0
```

Comment:
I think it is nice if we mention it is scheduled for 5.2.0 in [this](https://github.com/AxonFramework/AxonFramework/issues/3597) issue.